contracts/l2/FuseController.sol

```diff
@@ -0,0 +1,309 @@
+        address operator,
+        address from,
+        address to,
+        uint256 /*id*/,
+        uint256 value,
+        bytes calldata /*data*/,
+        bool operatorApproved
+    ) external view returns (bytes memory) {
+        TokenData memory td;
+
+        (
+            td.owner,
+            td.resolver,
+            td.expiry,
+            td.fuses,
+            td.renewalController
+        ) = _unpack(tokenData);
+
+        require(value == 1);
+        require(from == td.owner);
+        require(operator == td.owner || operatorApproved);
+
+        return
+            _pack(
+                address(this),
+                to,
+                td.resolver,
+                td.expiry,
+                td.fuses,
+                td.renewalController
+            );
+    }
+
+    function balanceOf(
+        bytes calldata tokenData,
+        address _owner,
+        uint256 /*id*/
+    ) external pure returns (uint256) {
+        (address owner, , , , ) = _unpack(tokenData);
+        return _owner == owner ? 1 : 0;
+    }
+
+    function resolverFor(
+        bytes calldata tokenData
+    ) external pure returns (address) {
+        (, address resolver, , , ) = _unpack(tokenData);
+        return resolver;
+    }
+
+    function expiryOf(bytes32 node) external view returns (uint64) {
+        // get the tokenData
+        bytes memory tokenData = registry.getData(uint256(node));
+        (, , uint64 expiry, , ) = _unpack(tokenData);
+        return expiry;
+    }
+
+    function fusesOf(bytes32 node) external view returns (uint96) {
+        // get the tokenData
+        bytes memory tokenData = registry.getData(uint256(node));
+        (, , , uint96 fuses, ) = _unpack(tokenData);
+        return fuses;
+    }
+
+    function renewalControllerOf(bytes32 node) external view returns (address) {
+        // get the tokenData
+        bytes memory tokenData = registry.getData(uint256(node));
+        (, , , , address renewalController) = _unpack(tokenData);
+        return renewalController;
+    }
+
+    function upgrade(bytes32 node, bytes calldata extraData) public {
+        // Make sure the upgrade contract is set.
+        if (address(upgradeContract) == address(0)) {
+            revert CannotUpgrade();
+        }
+
+        // Unpack the tokenData of the node.
+        bytes memory tokenData = registry.getData(uint256(node));
+        (
+            address owner,
+            address resolver,
+            uint64 expiry,
+            uint96 fuses,
+            address renewalController
+        ) = _unpack(tokenData);
+
+        bool isAuthorized = registry.getAuthorization(
+            uint256(node),
+            owner,
+            msg.sender
+        );
+
+        if (owner != msg.sender && !isAuthorized) {
+            revert Unauthorised(node, msg.sender);
+        }
+
+        if (_isExpired(tokenData)) {
+            revert nameExpired(node);
+        }
+
+        // Change the controller to the upgrade contract.
```

Comment:
You may want to instead allow the upgrade contract to specify the new data, so that a data format migration could take place at the same time as a contract upgrade.

Reply:
The upgrade target contract won't have permissions to make any changes if we don't change the controller here?

Reply:
Right, but we could have a function on the upgrade contract that we call with the old data, and it returns the new data to set on the registry.

Reply:
I don't have an opinion on this, could be either way.